src/main/java/io/supertokens/jwt/JWTSigningKey.java

```diff
@@ -0,0 +1,238 @@
+            } else {
+                keys = keysFromStorage;
+            }
+
+            return keys;
+        }
+
+        throw new QuitProgramException("Unsupported storage type detected");
+    }
+
+    /**
+     * Used to retrieve a key for JWT validation, for a given signing algorithm. If there are no keys in storage that match the given algorithm a new one is generated.
+     * @param algorithm The signing algorithm
+     * @return {@link JWTSigningKeyInfo} key
+     * @throws UnsupportedJWTSigningAlgorithmException If there is an error in the provided algorithm when getting/generating keys
+     * @throws StorageQueryException If there is an error interacting with the database
+     * @throws StorageTransactionLogicException If there is an error interacting with the database
+     */
+    public JWTSigningKeyInfo getOrCreateAndGetKeyForAlgorithm(SupportedAlgorithms algorithm)
+            throws UnsupportedJWTSigningAlgorithmException, StorageQueryException, StorageTransactionLogicException {
+        SessionStorage storage = StorageLayer.getSessionStorage(main);
+
+        if (storage.getType() == STORAGE_TYPE.SQL) {
+            SessionSQLStorage sqlStorage = (SessionSQLStorage) storage;
+
+            try {
+                return sqlStorage.startTransaction(con -> {
+                    JWTSigningKeyInfo keyInfo = null;
+
+                    List<JWTSigningKeyInfo> keysFromStorage = sqlStorage.getJWTSigningKeys_Transaction(con);
+
+                    // Loop through the keys and find the first one for the algorithm, if the list is empty a new key will be created after the loop
+                    for (int i = 0; i < keysFromStorage.size(); i++) {
+                        JWTSigningKeyInfo currentKey = keysFromStorage.get(i);
+                        if (algorithm.equalsString(currentKey.algorithm)) {
+                            keyInfo = currentKey;
+                            break;
+                        }
+                    }
+
+                    // If no key was found create a new one
+                    if (keyInfo == null) {
+                        keyInfo = generateAndSetKeyInStorage(sqlStorage, con, algorithm);
+                    }
+
+                    sqlStorage.commitTransaction(con);
+                    return keyInfo;
+                });
+            } catch (StorageTransactionLogicException e) {
+                 if (e.actualException instanceof UnsupportedJWTSigningAlgorithmException) {
+                    throw (UnsupportedJWTSigningAlgorithmException) e.actualException;
+                }
+
+                throw e;
+            }
+        } else if (storage.getType() == STORAGE_TYPE.NOSQL_1) {
+            SessionNoSQLStorage_1 noSQLStorage = (SessionNoSQLStorage_1) storage;
+
+            JWTSigningKeyInfo keyInfo = null;
+
+            while (true) {
+                List<JWTSigningKeyInfo> keysFromStorage = noSQLStorage.getJWTSigningKeys_Transaction();
```

Comment:
We should discuss how NoSQL logic works. It seems incorrect on a quick glance

Reply:
As discussed we will use findOneAndUpdate in the case of mongo for this